Add arithmetic and data management instructions

MMagueta opened the pull request.

MMagueta committed:
add arithmetic and data management instructions

EduardoLR10 commented on Chime/Instructions.m:
```diff
@@ -0,0 +1,80 @@
+    return;
+}
+
+void instruction_op_or(Stack *_dataStack){ NSInteger value1 = [[_dataStack pop] integerValue];
```
```suggestion
void instruction_op_or(Stack *_dataStack){ 
      NSInteger value1 = [[_dataStack pop] integerValue];
```
This amount of boilerplate made me think: Is there a way to generalize a binary function in C/ObjC without using `void*`?

EduardoLR10 commented on Chime/Instructions.m:
```diff
@@ -0,0 +1,80 @@
+    return;
+}
+
+void instruction_op_not(Stack *_dataStack){ NSInteger value = [[_dataStack pop] integerValue];
```
```suggestion
void instruction_op_not(Stack *_dataStack){ 
      NSInteger value = [[_dataStack pop] integerValue];
```

EduardoLR10 commented on Chime/Headers/Instructions.h:
```diff
@@ -0,0 +1,15 @@
+#import "Stack.h"
+#import <Foundation/Foundation.h>
+
+void instruction_op_fetch(NSMapTable* _registers, Stack *_dataStack, NSMutableArray *_ram);
+void instruction_op_load_a(NSMapTable* _registers, Stack *_dataStack, NSMutableArray *_ram);
+void instruction_op_store_a(NSMapTable* _registers, Stack *_dataStack, NSMutableArray *_ram);
+void instruction_op_and(Stack *_dataStack);
+void instruction_op_or(Stack *_dataStack);
+void instruction_op_not(Stack *_dataStack);
+void instruction_op_xor(Stack *_dataStack);
+void instruction_op_plus(Stack *_dataStack);
+void instruction_op_double(Stack *_dataStack);
+void instruction_op_half(Stack *_dataStack);
+void instruction_op_plus_star(Stack *_dataStack);
```
Is there any particular reason we should not make those methods of the VM?

z-silver replied:
We probably could, actually. I don't see any reason not to.

MMagueta replied:
Already done in https://github.com/Dr-Nekoma/chime/pull/7

z-silver commented on Chime/Instructions.m:
```diff
@@ -0,0 +1,80 @@
+}
+
+void instruction_op_not(Stack *_dataStack){ NSInteger value = [[_dataStack pop] integerValue];
+    [_dataStack push:@(!value)];
```
```suggestion
    [_dataStack push:@(~value)];
```
This was my bad, `!` is actually boolean negation. The bitwise not is actually `~`.

MMagueta replied:
Also solved in #7

z-silver committed:
Fix ISR mask

The correct bit pattern to mask off the bottom two bits is 100, with
the 1 being the bit used to extend the pattern to the left. That bit
pattern, interpreted as an integer and written in decimal, is -4.

z-silver committed:
Use a shift to express the number of bits to mask off

MMagueta committed:
add load and store for return stack and data stack

EduardoLR10 committed:
Change order of OP_HALT instruction in Opcode enum

EduardoLR10 committed:
Add printState function to check VM's internals

EduardoLR10 committed:
Fix implementation of NOT and alter indetation

EduardoLR10 committed:
Add implementation of printState

EduardoLR10 committed:
Add initilization for instruction stack

EduardoLR10 committed:
Add instruction numbers to help debugging

EduardoLR10 committed:
Fix errors with function call instruction

EduardoLR10 committed:
Fix indentation

EduardoLR10 committed:
Changing instruction functions to be a separate Category

EduardoLR10 committed:
Remove previous instructions method declarations

EduardoLR10 committed:
Add collectNextInstruction method

EduardoLR10 committed:
Add remaining instructions' methods implementation

EduardoLR10 committed:
Add assign and retain to properties of classes

EduardoLR10 committed:
Add justfile for Linux compilation

EduardoLR10 committed:
Simplify nix flake to only provide tooling for development

EduardoLR10 committed:
Add formatter to justfile

MMagueta committed:
indent Chime/Instructions.m

Co-authored-by: Eduardo Lemos Rocha <dudulr10@gmail.com>

MMagueta committed:
indent Chime/Instructions.m

Co-authored-by: Eduardo Lemos Rocha <dudulr10@gmail.com>

MMagueta committed:
fix boolean not operator

Co-authored-by: Mário Pinotti Jr. <marinho.pinotti@gmail.com>

MMagueta committed:
Merge branch 'master' into add-category-refactoring

MMagueta committed:
add framework foundation and update xcode proj

MMagueta committed:
remove chime from gitignore

MMagueta committed:
Merge branch 'add-arithmetic-data-management' into add-category-refactoring

MMagueta committed:
delete instructions

EduardoLR10 committed:
Merge pull request #7 from Dr-Nekoma/add-category-refactoring

Add Category + Refactoring

z-silver reviewed (changes requested):
Please add the directory `Chime.xcodeproj` to gitignore. Looks good otherwise.

EduardoLR10 committed:
Remove Chime.xcodeproj folder from upstream
